Support mapping volumes in Dev Service container based databases (#31859)
quarkusio/quarkus: docs/src/main/asciidoc/databases-dev-services.adoc

mcr.microsoft.com/mssql/server:2017-CU12
----

== Mapping volumes into Dev Services for Database

> Mapping volumes from our local system to the containers is handy to provide files from local to the database like scripts or configuration, but also to keep the whole database data that can be reused by a later execution.

Reviewer: ```suggestion
Mapping volumes from the Docker host's filesystem to the containers is handy to provide files like scripts or configuration, but also to preserve database data and reuse it after an application restart.
```

[NOTE]
====

> Mapping volumes will only work in Dev Services with a container based database like PostgreSQL. 

Reviewer: ```suggestion
Mapping volumes will only work in Dev Services with a container-based database like PostgreSQL. 
```

====

> Dev Services supports volumes from the filesystem and the classpath:

Reviewer: ```suggestion
Dev Services volumes can be mapped to the filesystem or the classpath:
```

<2> When using classpath volumes, the location has to start with "classpath:". The file or folder "./file" from the project's classpath will be accessible at "/container/to" in the container.

It's very important to remark that when using a classpath volume, the container will be granted only with read mode permission. On the other hand, when using a file system volume, the container will be granted with read and write permission. 

=== Example of mapping volumes to persist the database data

Let's see an example using PostgreSQL where we'll map a file system volume to keep the database data permantently and use it:

[source,properties]
----
quarkus.datasource.db-kind=postgresql
quarkus.datasource.devservices.volumes."/local/test/data"=/var/lib/postgresql/data
----

The container location where the data is placed depends on the database vendor, for PostgresSQL is "/var/lib/postgresql/data", but for MySQL, you would need this configuration instead:

[source,properties]
----
quarkus.datasource.db-kind=mysql
quarkus.datasource.devservices.volumes."/local/test/data"=/var/lib/mysql
----

> When starting the dev services (for example, in tests or in DEV mode), you will see that the folder "/local/test/data" will be created at your file sytem and that will contain all the database data. When rerunning again the same dev services, this data will contain all the data you might have created beforehand. 

Reviewer: ```suggestion
When starting Dev Services (for example, in tests or in DEV mode), you will see that the folder "/local/test/data" will be created at your file sytem and that will contain all the database data. When rerunning again the same dev services, this data will contain all the data you might have created beforehand. 
```

[IMPORTANT]
====

> When mapping volumes from your local file system to keep the database data, Quarkus might wipe out all the data if you have configured your application to drop and create the schema using `quarkus.hibernate-orm.database.generation = drop-and-create` (using `quarkus.hibernate-orm.database.generation=none` or `quarkus.hibernate-orm.database.generation=validate` is recommended). Also, using Flyway to migrate your schema when starting the application will modify the database data located at your local file system.

Reviewer: No, the data is at risk even if you didn't configure anything. That's the thing: for dev services, we default to wiping out the data...

```suggestion
When using Dev Services with Hibernate ORM, by default Quarkus will wipe out the database on application startup, which will wipe out the database data on your Docker host's filesystem.
Configure `quarkus.hibernate-orm.database.generation=none` or `quarkus.hibernate-orm.database.generation=validate` to avoid this behavior.

Also, using Flyway to migrate your schema when starting the application will modify the database data on your Docker hosts's file system.
```
